### PR TITLE
Hint TTS

### PR DESCRIPTION
**Pull request by mukulikadey**
The hint field now has the TTS (Text To Speech) functionality. By default the TTS is set to "English (US)". The user also has the option to disable the TTS in the settings of the deck.

### CODE REVIEW

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java [hunk 1]
```diff
@@ -199,6 +199,11 @@ public void onUpgrade(SQLiteDatabase database, ConnectionSource connectionSource
         if(oldVersion <= 7){
             database.execSQL(CREATE_MULTIPLE_CHOICE_TABLE);
         }
+
+        if(oldVersion <= 8){
+            database.execSQL("alter table settings add hintAudio String");
```
Comment (rekandiah) [inline app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: What are the values that both of the columns are holding? Why do we need two?

File: app/src/main/java/org/liberty/android/fantastischmemo/entity/Setting.java [hunk 2]
```diff
@@ -249,6 +255,12 @@ public boolean isAnswerAudioEnabled(){
         return !Strings.isNullOrEmpty(getAnswerAudio());
     }
 
+    public String getHintAudio(){return hintAudio;}
+
```
Comment (rekandiah) [inline app/src/main/java/org/liberty/android/fantastischmemo/entity/Setting.java]: fix spacing of curly brackets

Comment (rekandiah) [reply app/src/main/java/org/liberty/android/fantastischmemo/entity/Setting.java]: for all the functions you added on this file

File: app/src/main/java/org/liberty/android/fantastischmemo/utils/CardTTSUtil.java [hunk 3]
```diff
@@ -173,6 +196,18 @@ private void initTTS() {
         }  else {
             answerTTS = new NullAnyMemoTTS();
         }
+
+        if (setting.isHintAudioEnabled()){
+            String ha = setting.getHintAudio();
```
Comment (rekandiah) [inline app/src/main/java/org/liberty/android/fantastischmemo/utils/CardTTSUtil.java]: change the "ha" variable to something more meaningful

File: app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java (hunk 1 above)
Comment (mukulikadey) [reply app/src/main/java/org/liberty/android/fantastischmemo/common/AnyMemoDBOpenHelper.java]: hintAudio is the actual audio file, and hintAudioLocation is the path where all the audio files are located.

File: app/src/main/java/org/liberty/android/fantastischmemo/utils/CardTTSUtil.java (hunk 3 above)
Comment (mukulikadey) [reply app/src/main/java/org/liberty/android/fantastischmemo/utils/CardTTSUtil.java]: Lmao I know but they used "qa" (questionAudio) and "aa" (answerAudio), so I tried to keep it consistent with what they had before, hence the "ha" came about.

File: app/src/androidTest/java/org/liberty/android/fantastischmemo/test/db/SettingTest.java [hunk 4]
```diff
@@ -105,6 +105,15 @@ public void testAnswerAudio() throws Exception {
...
     }
 
+    @SmallTest
+    @Test
+    public void testHintAudio() throws Exception {
+        Setting setting = getCurrentSetting();
+        setting.setHintAudio("DE");
+        Setting savedSetting = setSetting(setting);
+        assertEquals("DE", savedSetting.getHintAudio());
+    }
```
Comment (coreylevine) [inline app/src/androidTest/java/org/liberty/android/fantastischmemo/test/db/SettingTest.java]: Is the audio sound set when the user adds a hint? Or do we have to set it after? What are the steps for setting a hint and then having the audio set?

File: app/src/main/java/org/liberty/android/fantastischmemo/ui/SettingsScreen.java [hunk 5]
```diff
@@ -300,6 +301,8 @@ public void run() {
 
             answerLocaleSpinner =  (AMSpinner)findViewById(R.id.answer_locale_spinner);
 
+            hintLocaleSpinner = (AMSpinner)findViewById(R.id.hint_locale_spinner);
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/ui/SettingsScreen.java]: Add space between ) and find...

File: app/src/main/java/org/liberty/android/fantastischmemo/utils/CardTTSUtil.java [hunk 6]
```diff
@@ -123,6 +141,11 @@ public void release() {
             answerTTS.destory();
             answerTTS = null;
         }
+
+        if(hintTTS != null){
```
Comment (coreylevine) [inline app/src/main/java/org/liberty/android/fantastischmemo/utils/CardTTSUtil.java]: Add space between ){

File: app/src/main/java/org/liberty/android/fantastischmemo/utils/CardTTSUtil.java (hunk 3 above)
Comment (coreylevine) [reply app/src/main/java/org/liberty/android/fantastischmemo/utils/CardTTSUtil.java]: Add space between ){

**Comment by j-suguma** [comment]
CI build for hintTTS:
image

tests:
image

static analysis:
[hintTTSSA.pdf](https://github.com/jameelbrochu/AnyMemo/files/1885382/hintTTSSA.pdf)

File: app/src/androidTest/java/org/liberty/android/fantastischmemo/test/db/SettingTest.java (hunk 4 above)
Comment (mukulikadey) [reply app/src/androidTest/java/org/liberty/android/fantastischmemo/test/db/SettingTest.java]: The hint audio by default is set to "English(US)", so it is always on from the start. However, the user also has the option to set the hint audio to "Disabled" in the settings of the deck, which turns off the TTS for hint. The same thing can be done to question and answer.

**Comment by coreylevine** [comment]
@mukulikadey put a description for your pull request please.

**Pull request merged**